### ISSUE TITLE
Allow for multiple `c.NotebookApp.notebook_dir` paths

### ISSUE THREAD

**Issue by mjbludwig** [jupyterlab/jupyterlab#9265]
Hi JupyterLab Team,

I don't want to beat a dead horse hear as this issue was brought up in ticket #2532, but I have a slightly different angle.

I deploy JupyterLab via the [Open OnDemand HPC portal](https://github.com/OSC/Open-OnDemand) system. This mechanism launches JupyterLab in a Per User NGINX process that is run under each users uid and privileges.

On the various clusters, there are various nfs, lustre, etc. filesystems that users have data in/work from. These are most often mounted off root such as `/scratch/users/[username]` or `/net/[nfs server]/username`. Currently, if I read the above mentioned ticket correct, to allow users to browse these directories I have to set `c.NotebookApp.notebook_dir="/"` and they can then navigate based from `/`. This is not a security concern as the JupyterLab instance is running under their user privileges and since users have ssh access to interactive/head-nodes, this navigation is the same as a normal ssh shell. 

My main reason for requesting the ability to specify multiple notebook dirs is so as not to overwhelm users when they are navigating the fs's. While `c.NotebookApp.notebook_dir="/"` works, there is no need for the users to see the entire root filesystem. 

I think the ideal solution would be to have `${HOME}` be set as the root notebook dir so the first option would be `/home/[USER]` in the file browser but then a dynamic list of other sections with their own file browser that each have admin specified paths. 

So if I specified an additional path like `/scratch/users/${USER}` to a variable like `c.NotebookApp.notebook_add_dir="/scratch/users/${USER}" or even multiple paths in a colon separated list it would populate the file browser like so:

 - /home/jsmith
    | python.py
    | readme.md
- /scratch/users/jsmith
    | big_dataset.nc
    | bigger_dataset.hdf
....

Then, those admin specified additional dirs could also have only sub directory access to there is no possibility for filesystem traversal.

Let me know if I can clarify anything here and many thanks!

- Morgan

**Comment by jasongrout** [comment]
Just to be clear - In your mind, are there multiple filebrowsers, one for each root, or one filebrowser with multiple top-level roots?

Technically, such a feature would require enhancing the server side (https://github.com/jupyter/jupyter_server/) to support multiple roots (either through making the existing ContentsManager more general, or having the server spawn one ContentsManager per root and somehow advertising that to the frontend). Then there would be the changes in jlab to either support multiple roots in the single filebrowser, or a change to create new filebrowsers for each root.

Another option is to create symbolic links for each user in their home directory pointing to these things. That solves things at the filesystem level, and JupyterLab/Jupyter Server will just automatically work.

**Comment by mjbludwig** [comment]
Thanks for your quick response @jasongrout !

Both ways would work, which ever one being easier to implement (if either one could be considered easy to implement...).

The symbolic link method does indeed work and is what I am suggesting to users now as only a few have needed this. This may become an issue down the road if increasing numbers of users come asking about it. Alternatively, we could create these links upon user creation but we do try to keep a barebones /home dir for new users as the level of *nix knowledge varies greatly. 

I understand this is really only an issue with HPC computing where /home directories are generally very small and most workflows are pointed to using large filesystem space for data storage and manipulation.

**Comment by jasongrout** [comment]
I think the easiest and best user interface is the symbolic links - that works in the ux today, and their JupyterLab filebrowser, local filebrowser, terminal, and other tools would all be consistent.

The next easiest would be separate filebrowsers. One way to push forward on that is:

A. Make Jupyter Server extension(s) for new contents managers, one manager for each root location. These provide the rest api that the JupyterLab filebrowser expects to communicate with.
B. Make a JupyterLab extension that creates a new filebrowser for each contents manager. JupyterLab already supports having multiple filebrowsers, each pointing to a different source, so this extension should be pretty straightforward, calling existing APIs and using existing JupyterLab components.

Hopefully that helps get someone started on exploring this.

**Comment by jasongrout** [comment]
You could also combine the server extension and JupyterLab extension into a single python package. The JLab 3.0 extension cookiecutter has a template to help someone get started with the boilerplate for a server extension and jlab extension in a single package.

**Comment by JohnOmernik** [comment]
So there is another angle on this:  Windows doesn't have a root. Therefore, the ability to specify multiple notebook_dir's is more critical.  If I have a user who wants to use their C: drives and their shared N: drive (or a network path like \\servername\share ) I either need to have multiple instances of jupyterlab or try to setup symlinks with mklink which is problematic because in some environments, symbolic links require administrative privs.   The ability to specify an array of notebook locations would be fantastic, and from here, that would be "the top level" in the file browser.  So if I did notebook_dirs = ['C:\notebooks', 'N:\shared_notebooks']. The filebrowser "top level" would show those two locations.  That would be fantastic.

**Comment by dhirschfeld** [comment]
> *Make a JupyterLab extension that creates a new filebrowser for each contents manager.*

Isn't this what [`jupyter-fs`](https://github.com/jpmorganchase/jupyter-fs) does?

**Comment by scruel** [comment]
Why not support multiple environment rather than just notebook paths(

**Comment by tchaton** [comment]
Yes, I am trying to figure out if JupyterLab has the same multi root workspaces functionalities as VSCode.

```code-workspace 
{ "folders": [ { "path": "/content" }, {"path": "/data"}, ], "settings": {} }
```

```python
code-server --bind-addr 0.0.0.0:8443  code-workspace
```

**Comment by tchaton** [comment]
> I think the easiest and best user interface is the symbolic links - that works in the ux today, and their JupyterLab filebrowser, local filebrowser, terminal, and other tools would all be consistent.
> 
> The next easiest would be separate filebrowsers. One way to push forward on that is:
> 
> A. Make Jupyter Server extension(s) for new contents managers, one manager for each root location. These provide the rest api that the JupyterLab filebrowser expects to communicate with. B. Make a JupyterLab extension that creates a new filebrowser for each contents manager. JupyterLab already supports having multiple filebrowsers, each pointing to a different source, so this extension should be pretty straightforward, calling existing APIs and using existing JupyterLab components.
> 
> Hopefully that helps get someone started on exploring this.

In some system, creating symbolic links isn't desired nor even possible. I think separate filebrowsers is the desired experience, similiar to VSCode.
